BUG: Fix groupby sorting on ordered Categoricals (GH25871)

kpflugshaupt opened the pull request:
As documented in #25871, groupby() on an ordered Categorical messes up category order when 'observed=True' is specified. 
Specifically, group labels will be ordered by first occurrence (as for an unordered Categorical), but grouped aggregation results will retain the Categorical's order.
The fix is a modified subset of #25173, which fixes a related case, but has not been merged yet.

- [x] closes #25871, also #25167
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

kpflugshaupt committed:
BUG: Fix groupby sorting on ordered Categoricals (GH25871) (#1)

* BUG: Fix groupby on ordered Categoricals (GH25871)

As documented in #25871, groupby() on an ordered Categorical messes up category order when 'observed=True' is specified. 
Specifically, group labels will be ordered by first occurrence (as for an unordered Categorical), but grouped aggregation results will retain the Categorical's order.
The fix is a modified subset of #25173, which fixes a related case, but has not been merged yet.

* BUG: Fix groupby on ordered Categoricals (GH25871)

* new test

* Fix groupby on ordered Categoricals (GH25871)

Testing all combinations of:
- ordered vs. unordered grouping column
- 'observed' True vs. False
- 'sort' True vs. False
In all cases, result group ordering must be correct. 
The test is built such that the result index labels are equal to aggregation results if all goes well (except for the one unobserved category)

kpflugshaupt committed:
BUG: GH25871 -- fix PEP 8 issues on test source

pep8speaks commented:
Hello @kpflugshaupt! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/groupby/test_categorical.py`](https://github.com/pandas-dev/pandas/blob/50e7d64f8a7be0848d32001611a8797bb8ef0453/pandas/tests/groupby/test_categorical.py):

> [Line 461:80](https://github.com/pandas-dev/pandas/blob/50e7d64f8a7be0848d32001611a8797bb8ef0453/pandas/tests/groupby/test_categorical.py#L461): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 461:84](https://github.com/pandas-dev/pandas/blob/50e7d64f8a7be0848d32001611a8797bb8ef0453/pandas/tests/groupby/test_categorical.py#L461): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 463:57](https://github.com/pandas-dev/pandas/blob/50e7d64f8a7be0848d32001611a8797bb8ef0453/pandas/tests/groupby/test_categorical.py#L463): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 464:61](https://github.com/pandas-dev/pandas/blob/50e7d64f8a7be0848d32001611a8797bb8ef0453/pandas/tests/groupby/test_categorical.py#L464): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 472:1](https://github.com/pandas-dev/pandas/blob/50e7d64f8a7be0848d32001611a8797bb8ef0453/pandas/tests/groupby/test_categorical.py#L472): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

kpflugshaupt committed:
BUG: GH25871 -- fix PEP 8 issues on test source

2nd shot...

kpflugshaupt commented:
Fixed new PEP 8 issues, too

kpflugshaupt commented:
Failing test reproduced here. Will investigate

kpflugshaupt committed:
BUG: GH25871 -- adapt test_groupby_levels_and_columns

This test had an adjustment for column order when 'observed=True' is set. This hid the fact that, with that parameter set, the data columns were not actually reordered -- it was just the column group labels (analogous to index labels in #25871), leaving the data columns in place and out of sync. (This was not visible as the data consisted only of ones).

I've made the test more sensitive (unsyncing of data columns will be caught now) and removed the special case for 'observed=True'. As there are no unobserved categories in this case, the result should not be influenced by this parameter.

kpflugshaupt commented:
Adapted the failing test, reopening

WillAyd commented:
@kpflugshaupt you don't need to close and reopen a PR for every commit (causes unnecessary messaging spam) - just push to the PR as you have new commits.

Is this a WIP or ready for review?

kpflugshaupt commented:
@WillAyd: Sorry for the spamming, I did not know. Will leave open in the future, thanks!
As soon as automatic checks come back OK, this is good for review.

gfyoung commented on pandas/tests/groupby/test_categorical.py:
```diff
@@ -453,6 +453,32 @@ def test_dataframe_categorical_with_nan(observed):
+        aggr[aggr.isna()] = 'AWOL'
+    tm.assert_equal(label, aggr)
+
+
```
Can you also add the example from #25167?  Then we can close that one as well (since they were duplicate issues AFAICT).

kpflugshaupt replied:
Will do, sure. I agree they seem duplicates.

kpflugshaupt replied:
Done, checking now.
I added the sample as-is, without extended parameterisation, as that was originally submitted.
My new test covers all combinations of parameters, anyway.
Is that what you had in mind?

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -347,7 +347,7 @@ Groupby/Resample/Rolling
 - Bug in :func:`pandas.core.groupby.GroupBy.first` and :func:`pandas.core.groupby.GroupBy.last` where timezone information would be dropped (:issue:`21603`)
 - Ensured that ordering of outputs in ``groupby`` aggregation functions is consistent across all versions of Python (:issue:`25692`)
-
+- Ensured that result group order is correct when grouping on an ordered Categorical and specifying ``observed=True`` (:issue:`25871`)
```
can you use double back-ticks around Categorical

kpflugshaupt replied:
Done

Commit:
Fix whatsnew file formatting

Commit:
Extend unit test with code sample from #25167

kpflugshaupt commented:
The added test from #25167 fails. Will investigate.

Commit:
Fix test: expected values

kpflugshaupt commented:
Got the expected values wrong, don't know what I was thinking. Fixed now.

Commit:
Fix test: expected values dtype (Windows takes 'int' as 'int32', instead of 'int64'), also PEP 8 fixes

kpflugshaupt commented:
Windows checks were failing because 'int' type apparently means 'int32' on that planet. I learned something today.

topper-123 commented:
> Windows checks were failing because 'int' type apparently means 'int32' on that planet. I learned something today.

This is a 32-bit issue, not Windows as such. This is a bit annoying, but pandas uses ``np.int64`` always in some cases and ``np.intp`` in other cases. In this case you should likely use ``np.intp``.

In general, you should in pandas always use explicit types, so either ``np.int64`` or ``np.intp``, depening on needs. Using just ``int`` causes various issues unless you actually want a python int.

kpflugshaupt commented:
That’s good to know. As I‘m working on Mac and Linux only, I assumed everyone defaulted to int64.

jreback commented on pandas/tests/groupby/test_categorical.py:
```diff
@@ -453,6 +453,44 @@ def test_dataframe_categorical_with_nan(observed):
+        aggr[aggr.isna()] = 'AWOL'
+    tm.assert_series_equal(label, aggr)
+
+    # GH 25167: Groupby with observed=True doesn't sort
```
this test doesn't use any of the parameters? can you update that. you *may* want to make it a new test (so don't have to jump thru hoops to create the expected result)

kpflugshaupt replied:
Can do. If I fully parameterise this test, it will overlap considerably with my new one, though, especially if I rewrite it along the lines of your other comment. As it is now, it's exactly the issue's sample code, thus proving that it's been solved.

But you can never have too many tests. I'll make it a separate one.

jreback replied:
really what am after here, IS this other test actually different? if not, then remove it and just add the issue number with the other. I agree can't have too many tests, but at the same times we ideally want *unique* tests.

kpflugshaupt replied:
I agree. Will remove the second test.

kpflugshaupt replied:
Removed in latest commit

jreback commented on pandas/tests/groupby/test_categorical.py:
```diff
@@ -453,6 +453,44 @@ def test_dataframe_categorical_with_nan(observed):
+    aggr = pd.Series(result.array)
+    if not observed:
+        aggr[aggr.isna()] = 'AWOL'
+    tm.assert_series_equal(label, aggr)
```
ideally would like this to simply create an expected result and compare the entire dataframe with assert_frame_equal. this might not be trivial to do though.

kpflugshaupt replied:
I had an earlier version, with only four parameter combinations, where I generated four expected DataFrames. This got old though, especially as the parameters cause subtle differences in the ordering of codes vs. categories in the result's Categorial index, even for the (visibly) same result data frame. 
In the end, I argued that for this issue, I'm really only interested in the result order anyway, rather than the internal details of the DataFrame. What's more, future refactoring might change those internals while leaving the result correctly ordered. 

Still, if you wish me to do it I will generate 8 expected result DataFrames to compare against. Your thoughts?

jreback replied:
i am kind of ok with the labels checking as you had done, though ideally would have hard coded expected results (which are then easy to see & logically interpret). even for the lables I would like to hard code them, or at least make them obvious.

kpflugshaupt replied:
Would you be OK with the check remaining the same, but the assertion reporting 
```
AssertionError: Labels and aggregation results not consistently sorted
for (ordered=True, observed=True, sort=True)
Result:
label
d    a
a    b
b    d
Name: val, dtype: object
```
in case of failure?
That should make things more obvious.

kpflugshaupt replied:
I've gone ahead and committed this, feel free to request further changes.

Commit:
Added #25167 to whatsnew file as resolved

kpflugshaupt commented:
That should be it -- all checks passing now. I've added #25167 to the solved issues in the whatsnew file.

kpflugshaupt committed:
Merge branch 'master' into master

kpflugshaupt committed:
Improve test reporting

Added a generic assert with a custom message to make problem more obvious.

kpflugshaupt committed:
Merge branch 'master' of https://github.com/kpflugshaupt/pandas

kpflugshaupt commented:
Changed test as per @jreback 's comments.

jreback commented:
thanks @kpflugshaupt nice patch!

kpflugshaupt commented:
You're welcome, nice working with you.
